### PR TITLE
docs(ruby): create Ruby on Rails section

### LINKED ISSUE #15365: Github security dependency PR doesn't actually upgrade the relevant package
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

Hello!

Renovate opened a PR in our repo to upgrade `actionpack` from 6.1.5 to 6.1.5.1, however that PR doesn't actually change that lib in the Gemfile.lock. The only library that changes is unrelated `loofah`. This https://github.com/advisories/GHSA-mm33-5vfq-3mm3 is the vulnerability. I'm not sure exactly why this happens, but I do see something odd in the logs. I'm posting them below.

### Relevant debug logs

<details><summary>Logs</summary>

Notice that `fixedVersion` below points to 6.1.5, shouldn't it be 6.1.5.1? I'm not sure this is the cause, but this is the only suspicious thing I could find in the logs.

```
DEBUG: processRepo()
DEBUG: Processing 3 branches: renovate_all, renovate_lock-file-maintenance, renovate_rubygems-actionpack-vulnerability
DEBUG: Calculated maximum PRs remaining this run
{
  "prsRemaining": 99
}
DEBUG: PullRequests limit = 99
DEBUG: Calculated maximum branches remaining this run
{
  "branchesRemaining": 99
}
DEBUG: Branches limit = 99
DEBUG: Setting current branch to master(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: latest commit(branch="renovate_rubygems-actionpack-vulnerability")
{
  "branchName": "master",
  "latestCommitDate": "2022-04-29T09:34:29+02:00"
}
DEBUG: getBranchPr(renovate_rubygems-actionpack-vulnerability)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: findPr(renovate_rubygems-actionpack-vulnerability, undefined, open)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: getPrList success(branch="renovate_rubygems-actionpack-vulnerability")
{
  "pullsTotal": 66,
  "requestsTotal": 1,
  "apiQuotaAffected": true
}
DEBUG: Found PR #6694(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: branchExists=true(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: PR rebase requested=true(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Checking if PR has been edited(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Branch has not been modified(branch="renovate_rubygems-actionpack-vulnerability")
{
  "branchName": "renovate_rubygems-actionpack-vulnerability"
}
DEBUG: Found existing branch PR(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Checking schedule(, null)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: No schedule defined(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Manual rebase requested via Dependency Dashboard(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Using reuseExistingBranch: false(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: manager.getUpdatedPackageFiles() reuseExistinbranch=false(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: bundler.updateLockedDependency: actionpack@6.1.5 -> 6.1.5.1 [Gemfile.lock](branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: bundler.updateArtifacts(Gemfile)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Using bundler version specified in lockfile(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Using ruby version specified in .ruby-version(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Using docker to execute(branch="renovate_rubygems-actionpack-vulnerability")
{
  "image": "ruby"
}
DEBUG: Found version constraint - checking for a compatible image to use(branch="renovate_rubygems-actionpack-vulnerability")
{
  "depName": "docker.io/renovate/ruby",
  "scheme": "ruby",
  "constraint": "2.6.2"
}
DEBUG: Found compatible image version(branch="renovate_rubygems-actionpack-vulnerability")
{
  "depName": "docker.io/renovate/ruby",
  "scheme": "ruby",
  "constraint": "2.6.2",
  "version": "2.6.2"
}
DEBUG: Resolved tag constraint(branch="renovate_rubygems-actionpack-vulnerability")
{
  "image": "docker.io/renovate/ruby",
  "tagConstraint": "2.6.2",
  "tagVersioning": "ruby",
  "tag": "2.6.2"
}
DEBUG: Fetching Docker image: docker.io/renovate/ruby:2.6.2(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Finished fetching Docker image(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Executing command(branch="renovate_rubygems-actionpack-vulnerability")
{
  "command": "docker run --rm --name=renovate_ruby --label=renovate_child -v \"/mnt/renovate/gh/ygt/spabreaks\":\"/mnt/renovate/gh/ygt/spabreaks\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e GEM_HOME -w \"/mnt/renovate/gh/ygt/spabreaks\" docker.io/renovate/ruby:2.6.2 bash -l -c \"install-tool bundler 1.17.3 && ruby --version && bundler lock --update actionpack\""
}
DEBUG: exec completed(branch="renovate_rubygems-actionpack-vulnerability")
{
  "cmd": "docker run --rm --name=renovate_ruby --label=renovate_child -v \"/mnt/renovate/gh/ygt/spabreaks\":\"/mnt/renovate/gh/ygt/spabreaks\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e GEM_HOME -w \"/mnt/renovate/gh/ygt/spabreaks\" docker.io/renovate/ruby:2.6.2 bash -l -c \"install-tool bundler 1.17.3 && ruby --version && bundler lock --update actionpack\"",
  "durationMs": 16666,
  "stdout": "Installing legacy tool bundler v1.17.3\nSuccessfully installed bundler-1.17.3\n1 gem installed\nBundler version 1.17.3\nruby 2.6.2p47 (2019-03-13 revision 67232) [x86_64-linux]\nFetching gem metadata from https://rubygems.org/.........\nFetching gem metadata from https://rubygems.org/.\nResolving dependencies........\nWriting lockfile to /mnt/renovate/gh/ygt/spabreaks/Gemfile.lock\n",
  "stderr": ""
}
DEBUG: Returning updated Gemfile.lock(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Updated 1 package files(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Getting updated lock files(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Writing package.json files(branch="renovate_rubygems-actionpack-vulnerability")
{
  "packageFiles": [
    "cypress/package.json",
    "webpack/package.json"
  ]
}
DEBUG: Writing any updated package files(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Updated 1 lock files(branch="renovate_rubygems-actionpack-vulnerability")
{
  "updatedArtifacts": [
    "Gemfile.lock"
  ]
}
DEBUG: Getting comments for #6694(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Found 0 comments(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: isBranchConflicted(master, renovate_rubygems-actionpack-vulnerability)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Setting git author name(branch="renovate_rubygems-actionpack-vulnerability")
{
  "gitAuthorName": "Renovate Bot"
}
DEBUG: Setting git author email(branch="renovate_rubygems-actionpack-vulnerability")
{
  "gitAuthorEmail": "bot@renovateapp.com"
}
DEBUG: 2 file(s) to commit(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Preparing files for committing to branch renovate_rubygems-actionpack-vulnerability(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: git commit(branch="renovate_rubygems-actionpack-vulnerability")
{
  "deletedFiles": [],
  "ignoredFiles": [],
  "result": {
    "author": null,
    "branch": "renovate_rubygems-actionpack-vulnerability",
    "commit": "7a7f41f134dbf2081b64a70f726a3bc6b5167dcd",
    "root": false,
    "summary": {
      "changes": 1,
      "insertions": 1,
      "deletions": 1
    }
  }
}
DEBUG: Pushing branch renovate_rubygems-actionpack-vulnerability(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: git push(branch="renovate_rubygems-actionpack-vulnerability")
{
  "result": {
    "pushed": [],
    "branch": {
      "local": "renovate_rubygems-actionpack-vulnerability",
      "remote": "renovate_rubygems-actionpack-vulnerability",
      "remoteName": "origin"
    },
    "ref": {
      "local": "refs/remotes/origin/renovate_rubygems-actionpack-vulnerability"
    },
    "remoteMessages": {
      "all": []
    }
  }
}
INFO: Branch updated(branch="renovate_rubygems-actionpack-vulnerability")
{
  "commitSha": "7a7f41f134dbf2081b64a70f726a3bc6b5167dcd"
}
DEBUG: Checking if we can automerge branch(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: mergeStatus=no automerge(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Ensuring PR(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: There are 0 errors and 0 warnings(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: getBranchPr(renovate_rubygems-actionpack-vulnerability)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: findPr(renovate_rubygems-actionpack-vulnerability, undefined, open)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Found PR #6694(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Found existing PR(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: Processing existing PR(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: updatePr(6694, Update dependency actionpack to v6.1.5.1 [SECURITY], body)(branch="renovate_rubygems-actionpack-vulnerability")
DEBUG: PR updated(branch="renovate_rubygems-actionpack-vulnerability")
{
  "pr": 6694
}
INFO: PR updated(branch="renovate_rubygems-actionpack-vulnerability")
{
  "pr": 6694,
  "prTitle": "Update dependency actionpack to v6.1.5.1 [SECURITY]"
}
DEBUG: PR is not configured for automerge(branch="renovate_rubygems-actionpack-vulnerability")

.
.
.

DEBUG: getBranchPr(renovate_rubygems-actionpack-vulnerability)
DEBUG: findPr(renovate_rubygems-actionpack-vulnerability, undefined, open)
DEBUG: Found PR #6694
DEBUG: Closing Dependency Dashboard
DEBUG: Retrieving issueList
DEBUG: Retrieved 1 issues
DEBUG: Removing any stale branches
DEBUG: config.repoIsOnboarded=true
DEBUG: Branch lists
{
  "branchList": [
    "renovate_all",
    "renovate_lock-file-maintenance",
    "renovate_rubygems-actionpack-vulnerability"
  ],
  "renovateBranches": [
    "renovate_rubygems-actionpack-vulnerability"
  ]
}
DEBUG: remainingBranches=
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by HonkingGoose**
## Changes

- Explain Renovate's problems with Ruby on Rails gems

## Context

Closes #15365.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: docs/usage/ruby.md [hunk 1]
```diff
@@ -25,3 +25,23 @@ If you always want to have the latest available version, consider specifying `ge
...
+
+```
+gem 'rails'
+gem `railties`
+gem 'actionmailer'
+gem 'actionpack'
+gem 'activemodel'
+gem 'activerecord'
+gem 'activesupport'
+```
```
Comment (HonkingGoose) [inline docs/usage/ruby.md:44]: Can a Ruby on Rails user please check if this example Gemfile makes sense?

File: docs/usage/ruby.md [hunk 2]
```diff
@@ -25,3 +25,23 @@ If you always want to have the latest available version, consider specifying `ge
...
+
+Running `bundler lock --update actionpack` won't do anything, apart from updating the `actionpack` gem.
```
Comment (HonkingGoose) [inline docs/usage/ruby.md:46]: This post https://github.com/renovatebot/renovate/issues/15365#issuecomment-1113912176 makes it sound like the real problem is with the `Gmefile.lock` not getting updated correctly when Renovate updates a single Rails gem?

Should we put that in the text? It feels like I'm missing some key part of the explanation...

**Review by HonkingGoose — COMMENTED** [review]
Some initial questions.

@DawidJanczak, can you check my work please?

File: docs/usage/ruby.md [hunk 3]
```diff
@@ -25,3 +25,23 @@ If you always want to have the latest available version, consider specifying `ge
...
+## Ruby on Rails
+
+Renovate doesn't support [Rails](https://rubyonrails.org/) gems correctly.
+This is because Rails's gems must be updated in a batch.
+
+For example, say you have these Rails gems in your Gemfile:
+
+```
+gem 'rails'
+gem `railties`
```
Comment (DawidJanczak) [inline docs/usage/ruby.md:38]: This one should have single quotes instead of backticks.

File: docs/usage/ruby.md [hunk 4]
```diff
@@ -25,3 +25,23 @@ If you always want to have the latest available version, consider specifying `ge
...
+
+## Ruby on Rails
+
+Renovate doesn't support [Rails](https://rubyonrails.org/) gems correctly.
+This is because Rails's gems must be updated in a batch.
+
+For example, say you have these Rails gems in your Gemfile:
+
+```
+gem 'rails'
```
Comment (DawidJanczak) [inline docs/usage/ruby.md:37]: This should be removed - `gem 'rails'` will require everything and in this particular scenario we don't want that (hence all the libraries specified below).

File: docs/usage/ruby.md (hunk 1 above)
Comment (DawidJanczak) [inline docs/usage/ruby.md:44]: We might want to add `gem 'bootsnap'` here. I'm not 100% sure it's necessary, but usual Rails setups have it.

**Review by DawidJanczak — COMMENTED** [review]
Thank you for picking this up @HonkingGoose , please let me know if there's still some unclear bits here.

File: docs/usage/ruby.md (hunk 3 above)
Comment (HonkingGoose) [reply docs/usage/ruby.md:38]:
```suggestion
gem 'railties'
```

I use backticks all the time, but here I should use single quotes! 🙈

File: docs/usage/ruby.md (hunk 4 above)
Comment (HonkingGoose) [reply docs/usage/ruby.md:37]:
```suggestion
```

File: docs/usage/ruby.md (hunk 1 above)
Comment (HonkingGoose) [reply docs/usage/ruby.md:44]: If that's the usual setup, let's put it in the example.

File: docs/usage/ruby.md [hunk 5]
```diff
@@ -25,3 +25,23 @@ If you always want to have the latest available version, consider specifying `ge
 
 You can install the [Renovate App](https://github.com/apps/renovate) on GitHub.
 Or you can check out [Renovate OSS](https://github.com/renovatebot/renovate) to self-host Renovate.
+
+## Ruby on Rails
+
+Renovate doesn't support [Rails](https://rubyonrails.org/) gems correctly.
```
Comment (HonkingGoose) [inline docs/usage/ruby.md:31]: > This is something that's very specific to Rails and as such I'm not sure if you want to handle it. If you do you'd essentially have to scan the `Gemfile` for all Rails gems (that's the ones mentioned here https://github.com/rails/rails#frameworks-and-libraries + `rails` + `railties`) and update all of them whenever a single one needs updating.

As Renovate won't work with Ruby on Rails gems, maybe we should explain how to disable Renovate for those gems? Or create our own `workarounds:ignoreRubyOnRailsUpdates` preset, with the full list of Rails gems?

```json
{
  "packageRules": [
    {
      "description": "Disable Renovate for any Ruby on Rails gems",
      "matchPackageNames": ["list", "all", "ruby", "on", "rails", "gems", "here"],
      "enabled": false
    }
  ]
}
```

Comment (DawidJanczak) [reply docs/usage/ruby.md:31]: Yeah adding a link to the docs on how to disable it for certain gems might be helpful here.

Just out of curiosity does renovate deal with these kinds of exceptions to the rules in other package managers at all?

Comment (HonkingGoose) [reply docs/usage/ruby.md:31]: I think we have two options:

1. Create and maintain a `workarounds:ignoreRubyOnRailsUpdates` preset with a list of Ruby on Rails gems that Renovate should ignore 
2. Put the example workaround in the docs and let our users create and maintain the list of Ruby on Rails gems

Which do you prefer @rarkins?

Comment (HonkingGoose) [reply docs/usage/ruby.md:31]: Another option to consider:

3. Create and maintain a `group:rubyOnRails` preset:
    ```json
    {
      "packageRules": [
        {
          "matchDatasources": [
            "rubygems"
          ],
          "matchPackageNames": [
            "list",
            "all",
            "Ruby on Rails packages",
            "here"
          ],
          "groupName": "Ruby on Rails packages"
        }
      ]
    }
    ```

Comment (HonkingGoose) [reply docs/usage/ruby.md:31]: @rarkins which option do you like best? Or maybe you have another idea?

**Comment by HonkingGoose** [comment]
Change of plans! Instead of documenting _why_ Renovate can't work with Ruby on Rails gems, lets fix the problem properly. 😉 

Follow issue:

- #17101

I'm closing this PR, as we no longer want the changes I'm making with it. 😄

File: docs/usage/ruby.md (hunk 5 above)
Comment (HonkingGoose) [reply docs/usage/ruby.md:31]: `@rarkins` has chosen option 3, see issue:

- #17101

**Pull request closed without merging**